services/autoscaling/tests/unit/test_modules_auto_scaling_computational.py

```diff
@@ -1029,3 +1030,159 @@ async def test_cluster_scaling_up_more_than_allowed_with_multiple_types_max_star
     )
+
+
+@pytest.mark.parametrize(
+    "dask_task_imposed_ec2_type, dask_ram, expected_ec2_type",
+    [
+        pytest.param(
+            None,
+            parse_obj_as(ByteSize, "128Gib"),
+            "r5n.4xlarge",
+            id="No explicit instance defined",
+        ),
+    ],
+)
+async def test_long_pending_ec2_is_detected_as_broken_terminated_and_restarted(
+    with_short_ec2_instances_max_start_time: EnvVarsDict,
+    minimal_configuration: None,
+    app_settings: ApplicationSettings,
+    initialized_app: FastAPI,
+    create_dask_task: Callable[[DaskTaskResources], distributed.Future],
+    ec2_client: EC2Client,
+    dask_task_imposed_ec2_type: InstanceTypeType | None,
+    dask_ram: ByteSize | None,
+    create_dask_task_resources: Callable[..., DaskTaskResources],
+    dask_spec_local_cluster: distributed.SpecCluster,
+    expected_ec2_type: InstanceTypeType,
+    mock_find_node_with_name_returns_none: mock.Mock,
+    mock_docker_tag_node: mock.Mock,
+    mock_rabbitmq_post_message: mock.Mock,
+    short_ec2_instance_max_start_time: datetime.timedelta,
+):
+    assert app_settings.AUTOSCALING_EC2_INSTANCES
+    assert (
+        short_ec2_instance_max_start_time
+        == app_settings.AUTOSCALING_EC2_INSTANCES.EC2_INSTANCES_MAX_START_TIME
+    )
+    # we have nothing running now
+    all_instances = await ec2_client.describe_instances()
+    assert not all_instances["Reservations"]
+    # create a task that needs more power
+    dask_future = await _create_task_with_resources(
+        ec2_client,
+        dask_task_imposed_ec2_type,
+        dask_ram,
+        create_dask_task_resources,
+        create_dask_task,
+    )
+    assert dask_future
+    # this should trigger a scaling up as we have no nodes
+    await auto_scale_cluster(
+        app=initialized_app, auto_scaling_mode=ComputationalAutoscaling()
+    )
+
+    # check the instance was started and we have exactly 1
+    instances = await _assert_ec2_instances(
+        ec2_client,
+        num_reservations=1,
+        num_instances=1,
+        instance_type=expected_ec2_type,
+        instance_state="running",
+    )
+
+    # as the new node is already running, but is not yet connected, hence not tagged and drained
+    mock_find_node_with_name_returns_none.assert_not_called()
+    mock_docker_tag_node.assert_not_called()
+    # check rabbit messages were sent
+    _assert_rabbit_autoscaling_message_sent(
+        mock_rabbitmq_post_message,
+        app_settings,
+        initialized_app,
+        dask_spec_local_cluster.scheduler_address,
+        instances_running=0,
+        instances_pending=1,
+    )
+    mock_rabbitmq_post_message.reset_mock()
+
+    assert instances
+    assert "LaunchTime" in instances[0]
+    original_instance_launch_time: datetime.datetime = deepcopy(
+        instances[0]["LaunchTime"]
+    )
+    await asyncio.sleep(1)  # NOTE: we wait here since AWS does not keep microseconds
+    now = arrow.utcnow().datetime
```

Comment:
I guess you could skip the sleep and simply mock the time `now = arrow.utcnow().datetime + 1 second somehow`? if desired...

Reply:
yes.